spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/pulsar/ConsumerConfigProperties.java

```diff
@@ -0,0 +1,540 @@
+/*
+ * Copyright 2023-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.autoconfigure.pulsar;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+
+import org.apache.pulsar.client.api.ConsumerCryptoFailureAction;
+import org.apache.pulsar.client.api.DeadLetterPolicy;
+import org.apache.pulsar.client.api.MessageId;
+import org.apache.pulsar.client.api.RegexSubscriptionMode;
+import org.apache.pulsar.client.api.SubscriptionInitialPosition;
+import org.apache.pulsar.client.api.SubscriptionMode;
+import org.apache.pulsar.client.api.SubscriptionType;
+
+import org.springframework.boot.context.properties.NestedConfigurationProperty;
+import org.springframework.boot.context.properties.PropertyMapper;
+import org.springframework.pulsar.core.ConsumerBuilderCustomizer;
+
+/**
+ * Configuration properties used to specify Pulsar consumers.
+ *
+ * @author Chris Bono
+ * @since 3.2
+ */
+public class ConsumerConfigProperties {
```

Comment:
Could this be nested within `PulsarProperties`?

Reply:
No, it is currently used by the SCSt Pulsar Binder config properties as well https://github.com/spring-cloud/spring-cloud-stream/blob/25af6a76b04598eb31bf2f84e59f45cee742b8ea/binders/pulsar-binder/spring-cloud-stream-binder-pulsar/src/main/java/org/springframework/cloud/stream/binder/pulsar/properties/PulsarBinderConfigurationProperties.java#L33-L46

Reply:
It shouldn't be doing that. From [the docs](https://docs.spring.io/spring-boot/docs/3.1.2/reference/htmlsingle/#features.external-config.typesafe-configuration-properties.java-bean-binding):

> The properties that map to `@ConfigurationProperties` classes available in Spring Boot, which are configured through properties files, YAML files, environment variables, and other mechanisms, are public API but the accessors (getters/setters) of the class itself are not meant to be used directly.

Please nest it and make the appropriate changes in SCSt Pulsar Binder.

Reply:
When the auto-config, binder, and core were in a single repo/project this was not an issue but I figured we would need to decouple it. I will do that and nest as you suggest.